Config restructure: distribute module configs via MQTT

hikinggrass opened the pull request:
Changes to reduce configuration parsing effort by parsing it once by the manager and then distribute it to the modules via MQTT.

This is achieved by first parsing the MQTT settings from the config, passing these to the modules whilst spawning them. The modules themselves then ask for their module config via MQTT, which is in turn provided to them from the manager.

schemas, manifests, types, interfaces, error definitions are published as retained topics that modules can access when needed.

hikinggrass committed:
WIP MQTT based config distribution (working)

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add ManagerSettings and reduce RuntimeSettings

get logging config and mqtt settings from command line or environment variables

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Merge fix

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Fix 3-tier mappings

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Only publish module provides, manifests and module config cache once

Retain these topics so the modules can get then if they need them

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Cleanup and documentation

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Address some compiler warnings

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Fix mqtt_server_port usage in everestjs

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
cleanup

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
cleanup

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Port everestrs to new MQTT based config loading mechanism

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Move MQTTSettings to mqtt abstraction

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Use MQTTSettings struct instead of individual parameters in Everest and MQTTAbstraction ctors

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Cleanup headers

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Fix config tests

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Enable code coverage

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Fix telemetry config

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
cleanup

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Update clang-format lint job to use v1.3.1 of run-clang-format

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Fix clang-format

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
cleanup

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Fix telemetry config (de)serialization

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Remove redundant manager flag

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
clang-format

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add tests for helper functions in module_config

Fix bug in has_extension function

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add additional Config tests

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Publish types as well so validation can work again

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Drop log level of some messages, fix typos

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
clang-format

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add user-config test

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Fixed test interface description

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add test for module that has an interface using types that implements a cmd and a var

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add simple config serialization test

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add test for config file in legacy json format

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add test for config file that does not exist

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Updated some doxygen comments

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Exit early if parse_command_line returned false

This is for example the case in the --help and --version parameters

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Log an error if a wrong module is started for a given module id

This can happen if the wrong binary is manually selected for a standalone module

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Some more clang-tidy fixes

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

a-w50 commented on CMakeLists.txt:
```diff
@@ -175,18 +182,21 @@ if((${CMAKE_PROJECT_NAME} STREQUAL ${PROJECT_NAME} OR ${PROJECT_NAME}_BUILD_TEST
-        DEPENDENCIES tests everest
+        BASE_DIRECTORY "${PROJECT_SOURCE_DIR}"
+        DEPENDENCIES ${PROJECT_NAME}_tests everest::framework
+        EXCLUDE "${CMAKE_SOURCE_DIR}/tests" "${CMAKE_BINARY_DIR}/*" "${CPM_SOURCE_CACHE}"
```
Why is the tests folder excluded?

a-w50 commented on everestpy/src/everest/misc.cpp:
```diff
@@ -14,26 +14,71 @@ static std::string get_ev_prefix_from_env() {
 }
 
-RuntimeSession::RuntimeSession() : RuntimeSession(get_ev_prefix_from_env(), get_ev_conf_file_from_env()) {
+static std::shared_ptr<Everest::MQTTSettings> get_mqtt_settings_from_env() {
```
this function should return by value, there is no need here (yet) to create a shared pointer

a-w50 commented on everestpy/src/everest/misc.cpp:
```diff
@@ -14,26 +14,71 @@ static std::string get_ev_prefix_from_env() {
 
-RuntimeSession::RuntimeSession(const std::string& prefix, const std::string& config_file) :
-    rs(std::make_shared<Everest::RuntimeSettings>(prefix, config_file)), config(create_config_instance(rs)) {
+static std::string get_ev_module_from_env() {
```
`get_ev_module_from_env` and `get_ev_log_conf_file_from_env` do the same.  I would suggest to write a common function, which gets the required env string per parameter

a-w50 commented on everestpy/src/everest/misc.cpp:
```diff
@@ -14,26 +14,71 @@ static std::string get_ev_prefix_from_env() {
 
-RuntimeSession::RuntimeSession() : RuntimeSession(get_ev_prefix_from_env(), get_ev_conf_file_from_env()) {
+static std::shared_ptr<Everest::MQTTSettings> get_mqtt_settings_from_env() {
+    const auto mqtt_everest_prefix = std::getenv("EV_MQTT_EVEREST_PREFIX");
```
I would move the hardcoded string literals like `EV_MQTT_EVEREST...`  to a common header as constants for better visibility.  Probably they should be also used in everestjs etc.

a-w50 commented on everestpy/src/everest/misc.cpp:
```diff
@@ -14,26 +14,71 @@ static std::string get_ev_prefix_from_env() {
+    const auto mqtt_external_prefix = std::getenv("EV_MQTT_EXTERNAL_PREFIX");
+    const auto mqtt_broker_socket_path = std::getenv("EV_MQTT_BROKER_SOCKET_PATH");
+    const auto mqtt_broker_host = std::getenv("EV_MQTT_BROKER_HOST");
+    const auto mqtt_broker_port = std::getenv("EV_MQTT_BROKER_PORT");
```
For better usability I probably would define only a single environment variable like:
`EV_MQTT_URI`, which could look like `socket://path_to_socket:0/prefix` or `tcp://localhost:1883/prefix` or something similar.  That way, tests don't need to set and know about the different env variables

a-w50 commented on everestpy/src/everest/misc.cpp:
```diff
@@ -14,26 +14,71 @@ static std::string get_ev_prefix_from_env() {
+    std::shared_ptr<Everest::MQTTSettings> mqtt_settings;
+    if (mqtt_broker_socket_path == nullptr) {
+        if (mqtt_broker_host == nullptr or mqtt_broker_port == nullptr) {
+            throw std::runtime_error("If EV_MQTT_BROKER_SOCKET_PATH is not set EV_MQTT_BROKER_HOST and "
```
The error message is really precise.

a-w50 commented on everestpy/src/everest/misc.cpp:
```diff
@@ -14,26 +14,71 @@ static std::string get_ev_prefix_from_env() {
+            EVLOG_warning << "Could not parse MQTT broker port, using default: " << mqtt_broker_port_;
+        }
+        mqtt_settings = std::make_shared<Everest::MQTTSettings>(mqtt_broker_host, mqtt_broker_port_,
+                                                                mqtt_everest_prefix, mqtt_external_prefix);
```
`mqtt_external_prefix` is not checked, would probably lead to undefined behaviour

a-w50 commented on everestpy/src/everest/misc.cpp:
```diff
@@ -14,26 +14,71 @@ static std::string get_ev_prefix_from_env() {
+        } catch (...) {
+            EVLOG_warning << "Could not parse MQTT broker port, using default: " << mqtt_broker_port_;
+        }
+        mqtt_settings = std::make_shared<Everest::MQTTSettings>(mqtt_broker_host, mqtt_broker_port_,
```
instead of declaring `mqtt_settings` beforehand, it could be just returned here directly

a-w50 commented on everestpy/src/everest/misc.cpp:
```diff
@@ -14,26 +14,71 @@ static std::string get_ev_prefix_from_env() {
+    this->mqtt_settings = get_mqtt_settings_from_env();
+
+    auto result = Everest::ModuleConfig::get_config(mqtt_settings, module_id);
+    this->rs = std::make_shared<Everest::RuntimeSettings>(result.at("settings"));
```
no need for shared pointer!

a-w50 commented on everestpy/src/everest/misc.cpp:
```diff
@@ -14,26 +14,71 @@ static std::string get_ev_prefix_from_env() {
+}
+
+RuntimeSession::RuntimeSession() {
+    auto module_id = get_ev_module_from_env();
```
const

a-w50 commented on everestpy/src/everest/misc.cpp:
```diff
@@ -14,26 +14,71 @@ static std::string get_ev_prefix_from_env() {
+
+    this->mqtt_settings = get_mqtt_settings_from_env();
+
+    auto result = Everest::ModuleConfig::get_config(mqtt_settings, module_id);
```
const

a-w50 commented on everestpy/src/everest/misc.hpp:
```diff
@@ -17,15 +17,18 @@ class RuntimeSession {
         return rs;
     }
 
+    std::shared_ptr<Everest::MQTTSettings> get_mqtt_settings() const {
```
no shared pointer! Just return const ref.
Also note that in this case function signature is misleading: it is declared as `const`, but it returns a non-const shared pointer to its member!

a-w50 commented on everestpy/src/everest/misc.hpp:
```diff
@@ -17,15 +17,18 @@ class RuntimeSession {
     }
 
 private:
     std::shared_ptr<Everest::RuntimeSettings> rs;
```
no shared pointer!

a-w50 commented on include/framework/everest.hpp:
```diff
@@ -48,9 +48,8 @@ struct ErrorFactory;
-            const std::string& mqtt_server_socket_path, const std::string& mqtt_server_address, int mqtt_server_port,
-            const std::string& mqtt_everest_prefix, const std::string& mqtt_external_prefix,
-            const std::string& telemetry_prefix, bool telemetry_enabled);
+            const std::shared_ptr<MQTTSettings> mqtt_settings, const std::string& telemetry_prefix,
```
`const std::shared_ptr` doesn't fit, `std::shared_ptr<const MQTTSettings>` probably would, but still, don't use shared pointer here, either pass by value or by const ref (the lifetime of the settings should be longer than the lifetime of the `Everest` instance I guess).

a-w50 commented on include/framework/runtime.hpp:
```diff
@@ -12,6 +12,7 @@
 #include <framework/ModuleAdapter.hpp>
 #include <sys/prctl.h>
```
include order is a bit messed up here, `sys/prctl.h` should go up and `framework` and `utils` belong togeter.  We should probably create an issue for fixing the include directory namespacing in everest framework

a-w50 commented on include/framework/runtime.hpp:
```diff
@@ -91,47 +92,65 @@ const auto TERMINAL_STYLE_ERROR = fmt::emphasis::bold | fg(fmt::terminal_color::
-    bool telemetry_enabled;
+/// \brief Settings needed by the manager to load and validate a config
+
+struct ManagerSettings {
```
It is a good idea to split up the manager settings and the runtime settings, but the way it is done here, it leads to some code duplication.  The `ManagerSettings` does also create the `RuntimeSettings`, so why not just make `RuntimeSettings` a part of `ManagerSettings`?

a-w50 commented on include/framework/runtime.hpp:
```diff
@@ -146,19 +165,24 @@ struct ModuleCallbacks {
-    std::string project_version;
-    std::string git_version;
+    std::string project_name;    ///< CMake project name
+    std::string project_version; ///< Human readable version number
```
Not sure if `Human` should be written in lower case as it is not a full sentence

a-w50 commented on include/framework/runtime.hpp:
```diff
@@ -146,19 +165,24 @@ struct ModuleCallbacks {
 class ModuleLoader {
 private:
     std::shared_ptr<RuntimeSettings> runtime_settings;
+    std::shared_ptr<MQTTSettings> mqtt_settings;
```
no shared pointer

a-w50 commented on include/utils/config.hpp:
```diff
@@ -256,21 +336,39 @@ class Config {
     ///
     /// \returns a set of object keys
-    static std::set<std::string> keys(json object);
+    static std::set<std::string> keys(const json& object);
```
should be a free function, not a static member function
same goes for all the other static member functions

hikinggrass replied:
lets refactor this in a followup PR

a-w50 commented on include/utils/config.hpp:
```diff
@@ -43,50 +46,27 @@ struct schemas {
 
 ///
-/// \brief Contains config and manifest parsing
+/// \brief Base class for configs
```
*inheritance is the base class of evil* (https://www.youtube.com/watch?v=2bLkxj6EVoM)
This doesn't look right to me.  I think this should be solved using composition

a-w50 commented on include/utils/config.hpp:
```diff
@@ -203,34 +178,139 @@ class Config {
+    /// \brief checks if the given \p module_id provides the requirement given in \p requirement_id
+    ///
+    /// \returns a json object that contains the requirement
+    json resolve_requirement(const std::string& module_id, const std::string& requirement_id) const;
```
Why for example should a module need to resolve a requirement.  This is clearly the managers responsibility

a-w50 commented on include/utils/config_cache.hpp:
```diff
@@ -18,5 +18,17 @@ struct ConfigCache {
+        c.cmds = j.at("cmds");
+    }
+};
+NLOHMANN_JSON_NAMESPACE_END
```
btw, these functions should be forward declared and implemented in a cpp file.

a-w50 commented on include/utils/module_config.hpp:
```diff
@@ -0,0 +1,49 @@
+bool has_extension(const std::string& path, const std::string& ext);
+
+/// \returns a path that has been prefixed by \p prefix from the provided json \p value
+std::string get_prefixed_path_from_json(const nlohmann::json& value, const fs::path& prefix);
```
These functions are not really related to `module_config`, they should be moved to a *more* common header file

a-w50 commented on include/utils/module_config.hpp:
```diff
@@ -0,0 +1,49 @@
+    ModuleConfig() = default;
+
+    /// \brief get config from manager via mqtt
+    static json get_config(std::shared_ptr<MQTTSettings> mqtt_settings, const std::string& module_id);
```
This class has only one static function.  So it doesn't make sense to have this class at all.  Just make a free function out of `get_config`.

a-w50 commented on include/utils/module_config.hpp:
```diff
@@ -0,0 +1,49 @@
+
+#include <nlohmann/json.hpp>
+
+#include <utils/mqtt_abstraction.hpp>
```
Why should `module_config` need to know about the `mqtt_abstraction`?

a-w50 commented on include/utils/module_config.hpp:
```diff
@@ -0,0 +1,49 @@
+bool has_extension(const std::string& path, const std::string& ext);
+
+/// \returns a path that has been prefixed by \p prefix from the provided json \p value
+std::string get_prefixed_path_from_json(const nlohmann::json& value, const fs::path& prefix);
```
consistency, here you use `nlohmann::json`, above just `json`.  In general it is a good idea to use full namespaces in header files and use `using` for shortening (in contrast to giving meaning) in the cpp files

hikinggrass replied:
I've reduced the usage of this in multiple headers

a-w50 commented on include/utils/mqtt_abstraction.hpp:
```diff
@@ -15,14 +15,34 @@ using json = nlohmann::json;
 class MQTTAbstractionImpl;
 
+/// \brief minimal MQTT connection settings needed for an initial connection of a module to the manager
+struct MQTTSettings {
```
This class clearly belongs into its own header file, this will also solve the problem, mentioned above.
Mqtt settings don't need to be aware of the interface how you access the mqtt bus.

a-w50 commented on include/utils/mqtt_abstraction_impl.hpp:
```diff
@@ -59,15 +59,15 @@ class MQTTAbstractionImpl {
     ///
     /// \brief publishes the given \p data on the given \p topic with the given \p qos
-    void publish(const std::string& topic, const std::string& data, QOS qos);
+    void publish(const std::string& topic, const std::string& data, QOS qos, bool retain = false);
```
This `retain` behavior better might be in a different commit

a-w50 commented on include/utils/types.hpp:
```diff
@@ -39,6 +39,7 @@ enum class HandlerType {
     SubscribeVar,
     SubscribeError,
     ClearErrorRequest,
+    Internal,
```
`Internal` is really broad, could be everything.

a-w50 commented on include/utils/types.hpp:
```diff
@@ -82,6 +83,8 @@ struct ModuleInfo {
 
 struct TelemetryConfig {
     int id;
+    explicit TelemetryConfig(int id) : id(id) {
+    }
```
standard list initialization should be enough.  This struct is also a bit redundant.  Something like `using TelemetryConfigID = int` would probably have been enough

a-w50 commented on include/utils/types.hpp:
```diff
@@ -138,6 +141,44 @@ template <> struct adl_serializer<Mapping> {
+        }
+        return m;
+    }
+};
```
implementation should go into a cpp file

a-w50 commented on lib/module_config.cpp:
```diff
@@ -0,0 +1,159 @@
+        res_promise.set_value(std::move(data));
+    };
+
+    std::shared_ptr<TypedHandler> res_token =
```
no shared pointer

a-w50 commented on lib/module_config.cpp:
```diff
@@ -0,0 +1,159 @@
+}
+
+json ModuleConfig::get_config(std::shared_ptr<MQTTSettings> mqtt_settings, const std::string& module_id) {
+    auto mqtt = MQTTAbstraction(mqtt_settings);
```
It should not be the `ModuleConfig`s responsibility to spawn an mqtt client thread in order to retreive its configuration.  The mqtt client will be necessary later on anyway for communication, why not use that instance to retreive the configuration?

a-w50 commented on lib/module_config.cpp:
```diff
@@ -0,0 +1,159 @@
+    result["error_map"] = error_types_map;
+
+    auto module_config_cache_topic = fmt::format("{}module_config_cache", mqtt_settings->mqtt_everest_prefix);
+    auto module_config_cache = mqtt.get(module_config_cache_topic, QOS::QOS2);
```
const

a-w50 commented on lib/module_config.cpp:
```diff
@@ -0,0 +1,159 @@
+    } else if (res_future_status == std::future_status::ready) {
+        EVLOG_verbose << "res future ready";
+        result = res_future.get();
+    }
```
If having an `if` containing a `throw` , you don't need to do an `else if`.  Also a log like `res future ready` will probably help no one.

a-w50 commented on lib/config.cpp:
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: Apache-2.0
-// Copyright 2020 - 2023 Pionix GmbH and Contributors to EVerest
+// Copyright Pionix GmbH and Contributors to EVerest
```
This file is just to much for me to reason about.  I think the chosen abstraction should be improved and a clear statement about the responsibilities of `ManagerConfig`, `ConfigBase`, `Config` and `ModuleConfig`   should be done for  in the documentation

a-w50 commented on lib/config.cpp:
```diff
@@ -790,33 +852,63 @@ ModuleConfigs Config::get_module_configs(const std::string& module_id) const {
 }
 
-const json& Config::get_manifests() {
+const json& ConfigBase::get_manifests() {
```
function should be const

a-w50 commented on include/utils/error/error_type_map.hpp:
```diff
@@ -50,6 +56,12 @@ class ErrorTypeMap {
+    /// \brief Returns the contained ErrorType map
+    /// \return The error types map
+    ///
+    std::map<ErrorType, std::string> get_error_types_map();
```
naming!  `ErrorTypeMap` is a map of probably multiple error types, therefore it is okay to name the private property `error_types`.  But then the function can't be called `get_error_types_map`, either `get_error_types` or `get_error_type_map` (but for the latter one you could also just write `get` as the class is already called `ErrorTypeMap`.  Also note that with this function you expose the full private data of this class, rendering it useless with respect to *privacy* or *implementation details*.  This whole thing probably just wants to be a `using ErrorTypeMap = std::map<ErrorType, std::string>` with free functions like
```
ErrorTypeMap load_error_types(..);
/* load_error_types_map will not be needed */
std::string get_description(const ErrorTypeMap&, const ErrorType&);
bool has(const ErrorTypeMap&, const ErrorType&);
/* get_error_types_map is not needed */
```

hikinggrass replied:
Renamed the funtion to get_error_types. Maybe this can be refactored further in the future

a-w50 commented on include/utils/module_config.hpp:
```diff
@@ -0,0 +1,49 @@
+using json = nlohmann::json;
+namespace fs = std::filesystem;
+
+struct BootException : public std::runtime_error {
```
why is `BootException` inside `module_config`?  Either namespace it or prefix it which `Module` because catching `BootException` somewhere in the code will not make easily clear that this belongs to the startup of a module.

a-w50 commented on include/utils/mqtt_abstraction.hpp:
```diff
@@ -15,14 +15,34 @@ using json = nlohmann::json;
+    int mqtt_broker_port = 0;            ///< The port the MQTT broker listens on
+    std::string mqtt_everest_prefix;     ///< MQTT topic prefix for the "everest" topic
+    std::string mqtt_external_prefix;    ///< MQTT topic prefix for external topics
+    bool socket = false; ///< Indicates if a Unix Domain Socket is used for connection to the MQTT broker
```
this could probably be a function, checking if `mqtt_broker_socket_path` is not empty.  Furthermore, I wouldn't prefix all the properties with mqtt, because the struct is already called `MqttSettings`

a-w50 commented on include/utils/mqtt_abstraction.hpp:
```diff
@@ -15,14 +15,34 @@ using json = nlohmann::json;
-                    const std::string& mqtt_server_port, const std::string& mqtt_everest_prefix,
-                    const std::string& mqtt_external_prefix);
+    /// \brief Create a MQTTAbstraction with the provideded \p mqtt_settings
+    MQTTAbstraction(const std::shared_ptr<MQTTSettings> mqtt_settings);
```
no shared pointer

a-w50 commented on lib/everest.cpp:
```diff
@@ -33,17 +33,15 @@ const auto remote_cmd_res_timeout_seconds = 300;
-                 const std::string& telemetry_prefix, bool telemetry_enabled) :
-    mqtt_abstraction(mqtt_server_socket_path, mqtt_server_address, std::to_string(mqtt_server_port),
-                     mqtt_everest_prefix, mqtt_external_prefix),
+                 const std::shared_ptr<MQTTSettings> mqtt_settings, const std::string& telemetry_prefix,
```
no shared pointers!

a-w50 commented on lib/everest.cpp:
```diff
@@ -520,11 +519,12 @@ void Everest::subscribe_error(const Requirement& req, const error::ErrorType& er
     // check if requirement is allowed to publish this error_type
     // split error_type at '/'
-    int pos = error_type.find('/');
+    size_t pos = error_type.find('/');
```
just use `const auto`

a-w50 commented on lib/mqtt_abstraction.cpp:
```diff
@@ -1,21 +1,38 @@
+        mqtt_abstraction = std::make_unique<MQTTAbstractionImpl>(
+            mqtt_settings->mqtt_broker_host, std::to_string(mqtt_settings->mqtt_broker_port),
+            mqtt_settings->mqtt_everest_prefix, mqtt_settings->mqtt_external_prefix);
     }
```
always try to prefer direct initialization.  Here you can just write a simple factory method like
```
auto create_mqtt_client(const MQTTSettings& sessings) {
  if (settings.socket) {
    return std::make_unique<...>(..);
  } else {
    return std::make_unique<...>(..);
  }
}
```
And then the constructor will look like
```
MQTTAbstraction::MQTTAbstraction(const MQTTSettings& settings) : mqtt_abstraction(create_mqtt_client(settings)) {}
```
Note that the helper function `create_mqtt_client` can also be easily reused in tests and also properly tested by itself!

a-w50 commented on lib/mqtt_abstraction_impl.cpp:
```diff
@@ -198,6 +202,39 @@ void MQTTAbstractionImpl::unsubscribe(const std::string& topic) {
+
+    Handler res_handler = [this, &res_promise](json data) { res_promise.set_value(std::move(data)); };
+
+    std::shared_ptr<TypedHandler> res_token =
```
no shared pointer!

a-w50 commented on lib/mqtt_abstraction_impl.cpp:
```diff
@@ -198,6 +202,39 @@ void MQTTAbstractionImpl::unsubscribe(const std::string& topic) {
+
+    // wait for result future
+    std::chrono::time_point<std::chrono::steady_clock> res_wait =
+        std::chrono::steady_clock::now() + std::chrono::seconds(10);
```
hardcoded value -> explicit constant

a-w50 commented on lib/runtime.cpp:
```diff
@@ -348,6 +329,26 @@ RuntimeSettings::RuntimeSettings(const std::string& prefix_, const std::string&
+
+    run_as_user = settings.value("run_as_user", "");
+
+    auto version_information_path = data_dir / "version_information.txt";
```
hardcoded string literal -> explicit constant

a-w50 commented on lib/runtime.cpp:
```diff
@@ -269,6 +244,12 @@ RuntimeSettings::RuntimeSettings(const std::string& prefix_, const std::string&
         controller_rpc_timeout_ms = defaults::CONTROLLER_RPC_TIMEOUT_MS;
     }
 
+    std::string mqtt_broker_socket_path;
+    std::string mqtt_broker_host;
+    int mqtt_broker_port = 0;
+    std::string mqtt_everest_prefix;
+    std::string mqtt_external_prefix;
+
```
what is this for?

a-w50 commented on lib/runtime.cpp:
```diff
@@ -392,29 +390,53 @@ ModuleCallbacks::ModuleCallbacks(const std::function<void(ModuleAdapter module_a
 }
 
 int ModuleLoader::initialize() {
+    if (this->should_exit) {
```
If one needs to check for `should_exit` in an `initialize` function, probably the class' construction is not managed in a proper way.

a-w50 commented on lib/runtime.cpp:
```diff
@@ -563,9 +598,97 @@ bool ModuleLoader::parse_command_line(int argc, char* argv[]) {
+
+    // overwrite mqtt broker port with environment variable
+    // NOLINTNEXTLINE(concurrency-mt-unsafe): not problematic that this function is not threadsafe here
+    const char* mqtt_server_port = std::getenv("MQTT_SERVER_PORT");
```
isn't this code duplicate to the one in the everestpy.cpp for example?

a-w50 commented on src/manager.cpp:
```diff
@@ -237,26 +274,110 @@ static std::map<pid_t, std::string> spawn_modules(const std::vector<ModuleStartI
+                             QOS::QOS2, true);
+
+    auto module_config_cache = config.get_module_config_cache();
+    mqtt_abstraction.publish(fmt::format("{}module_config_cache", ms->mqtt_settings->mqtt_everest_prefix),
```
What are these topics for?  Who is going to use them?

a-w50 commented on tests/test_config.cpp:
```diff
@@ -8,99 +8,206 @@
+        }
+    }
+    GIVEN("A valid config with a valid module and enabled schema validation") {
+        std::shared_ptr<Everest::ManagerSettings> rs =
```
use auto

hikinggrass committed:
Update clang-tidy config

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Remove duplicate code

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Get MQTT everest and external prefixes or fallback to a default

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
const

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Change shared_ptr to const ref for MQTTSettings

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Do not exclude tests dir in gcovr coverage

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Keep lcov coverage in sync with gcovr coverage setup

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Directly return MQTTSettings

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Change shared_ptr to const ref for RuntimeSettings

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Cleanup include order

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add RuntimeSettings as member to ManagerSettings to avoid code duplication

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Use auto in tests

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Move to_json and from_json implementations from headers

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Move filesystem helper functions into their own header

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Move BootException to the types header

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Remove ModuleConfig class and replace move its get_config member to a get_module_config function

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Rename Internal handler to GetConfig

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Remove redundant "mqtt_" prefix from MQTTSettings members

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add getters for everest and external mqtt prefixes

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Use a shared MQTTAbstraction for config getting and EVerest afterwards

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Make more variables const

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
clang-format

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add misc-const-correctness check to clang-tidy config

Reorder checks so enabled checks come first, followed by checks that are deactivated

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Make more variables const

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Use std::size_t to avoid narrowing conversions

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Use int64_t instead of int in return value to avoid narrowing conversion

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Use constexpr instead of define for MQTT_BUF_SIZE

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Remove unused variables

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Cleanup

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Pass ManagerSettings as const ref instead of shared_ptr

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Fix config not being checked in check_config command

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Move module startup related environment variable handling to common function

Move environment variable names to constants

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Move version_information.txt filename to constant

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Move MQTT get timeout to constant

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
cleanup

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Use std::size_t everywhere

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Move parse_3_tier_model_mapping to ManagerConfig

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
re-order member ConfigBase, ManagerConfig and Config implementations

this now better reflects the order from the config header file that got shuffled around during breaking up into different classes

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Fix raw pointer usage

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add ImplementationInfo struct replacing json

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Merge remote-tracking branch 'origin/main' into feature/config-restructure

Fix build for C++, JS and PY bindings. TODO: RS

# Conflicts:
#	CMakeLists.txt
#	everestrs/everestrs/src/everestrs_sys.cpp
#	everestrs/everestrs/src/everestrs_sys.hpp
#	everestrs/everestrs/src/lib.rs
#	include/utils/config.hpp
#	include/utils/types.hpp
#	lib/config.cpp
#	lib/runtime.cpp
#	lib/types.cpp

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add build* to .gitignore

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Fix everestrs build again

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
clang-format

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
clang-format

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
reorder config.cpp

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Merge remote-tracking branch 'origin/main' into feature/config-restructure

# Conflicts:
#	everestrs/everestrs/src/everestrs_sys.cpp
#	everestrs/everestrs/src/everestrs_sys.hpp

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
typo

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Use get_module_name instead of extract_implementation_info to access module name

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Make extract_implementation_info into a free function

Add create_printable_identifier free function and use it in the member function

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Move to/from json impl of ConfigCache to cpp file

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
cleanup

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Move MQTTSettings into its own files

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add missing newlines

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
More const and .at() usage

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Use factory method create_mqtt_client to initialize MQTTAbstractionImpl based on MQTTSettings

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
cleanup

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Use ms in get timeout

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
unregister handler for mqtt get before throwing TimeOutException

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Use Add mqtt get config timeout ms

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
extract_implementation_info doesn't need access to config

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Split type definition publishing up into individual topics

This is identical to how it's done for interfaces to keep message sizes as small as possible

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Remove lcov coverage, add gcovr xml coverage report

Move this configuration to tests/CMakeLists.txt

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
more const

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Replace socket member with uses_socket() function

This checks if broker_socket_path is set

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
add missing newline

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Reduce usage of using for nlohmann::json in headers

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
clang-format

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Merge remote-tracking branch 'origin/main' into feature/config-restructure

# Conflicts:
#	include/utils/config.hpp
#	lib/config.cpp
#	lib/everest.cpp

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Re-introduce functionality to RuntimeSession ctor with config file param

This prevents everest-testing from breaking until it is refactored to use the new ctor + environment variables

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Use uses_socket() in manager instead of manually checking

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
rename get_error_types function

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
cleanup comments

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
turn shared_ptr into unique_ptr

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

a-w50 committed:
refactor(tests): dry

Signed-off-by: aw <aw@pionix.de>

hikinggrass committed:
Updated catch2 to latest stable release 3.7.1

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Move testing logic to tests subdir

Set coverage flags directly on the appropriate targets
To make this work from the tests directory set CMake policy CMP0079

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Re-use the config that's already loaded in the ManagerSettings

Don't load the same yaml file twice

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Merge remote-tracking branch 'origin/main' into feature/config-restructure

# Conflicts:
#	include/utils/message_queue.hpp
#	lib/everest.cpp
#	lib/message_queue.cpp
#	lib/mqtt_abstraction_impl.cpp
#	src/manager.cpp

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
clang-format

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Bump version to 0.19.0

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add documentation with sequence and class diagrams about MQTT config feature

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
fix typo

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
fix some small codacy issues in markdown

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

a-w50 commented on CMakeLists.txt:
```diff
@@ -30,9 +30,11 @@ ev_setup_python_executable(
+    if(POLICY CMP0079)
+        cmake_policy(SET CMP0079 NEW)
+        set(CMAKE_POLICY_DEFAULT_CMP0079 NEW)
+    endif()
```
We shouldn't need this, as the required cmake version is already > 3.13.
Furthermore, `set(CMAKE_POLICY_DEFAULT_CMP0079 NEW)` already enforces `cmake_policy(SET CMP0079 NEW)`, so only the latter would be necessary.  I also don't get the point of `if(POLICY CMP0079)`

hikinggrass replied:
I'm also very confused why this is needed, since it should be the default. However on CMake 3.31 I get a
```
Attempt to add link library "gcov" to target "manager" which is not built
  in this directory.

  This is allowed only when policy CMP0079 is set to NEW.
Call Stack (most recent call first):
  tests/CMakeLists.txt:69 (append_coverage_compiler_flags_to_target)
```
error as if that policy isn't set.
the `if(POLICY CMP0079)` just ensures that the policy is available before setting it, maybe not necessary on CMake >= 3.13

a-w50 commented on lib/everest.cpp:
```diff
@@ -42,7 +42,7 @@ Everest::Everest(std::string module_id_, const Config& config_, bool validate_da
                  std::shared_ptr<MQTTAbstraction> mqtt_abstraction, const std::string& telemetry_prefix,
                  bool telemetry_enabled) :
     mqtt_abstraction(mqtt_abstraction),
-    config(std::move(config_)),
```
moving a `const` is indeed pointless.  But it might be a better choice to make the constructor argument `Config config_` instead of const ref and dropping the move.

a-w50 commented on lib/everest.cpp:
```diff
@@ -188,8 +188,8 @@ Everest::Everest(std::string module_id_, const Config& config_, bool validate_da
 
     // register handler for global ready signal
-    const Handler handle_ready_wrapper = [this](json data) { this->handle_ready(data); };
-    const std::shared_ptr<TypedHandler> everest_ready =
```
why have all the `const`s disappeared?  `auto` can also be used for the `shared_ptr` variables

hikinggrass replied:
Looks like that's from the recent merge of main where this turned from `Handler` to `auto`. I'll change that

hikinggrass committed:
Remove redundant cmake_policy command

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Re-introduce accidentally removed const

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>
